parla/task_runtime.py

```diff
@@ -13,74 +13,63 @@
+from parla.environments import EnvironmentComponentInstance, TaskEnvironmentRegistry, TaskEnvironment
 from parla.cpu_impl import cpu
+from parla.dataflow import Dataflow
 
 # Logger configuration (uncomment and adjust level if needed)
 #logging.basicConfig(level = logging.DEBUG)
 logger = logging.getLogger(__name__)
 
-
 __all__ = ["Task", "SchedulerContext", "DeviceSetRequirements", "OptionsRequirements", "ResourceRequirements", "get_current_devices"]
 
-
 # Note: tasks can be implemented as lock free, however, atomics aren't really a thing in Python, so instead
 # make each task have its own lock to mimic atomic-like counters for dependency tracking.
 
-
 TaskAwaitTasks = namedtuple("AwaitTasks", ("dependencies", "value_task"))
 
 
-class TaskID:
-    pass
-
-
-class WaitDependeesCollection:
-    """ If a task has dependent tasks which are not spawned yet,
-        this class stores the task until all of them are spawned.
-        This class exploits dictionary of TaskID key for a task
-        which is not yet created and makes dependees wait,
-        and a list of TaskID value, which is a list of tasks
-        which are dependees of and are waiting for the task
-        corresponding to the key.
+class UnspawnedDependencies:
+    """ Collection of dependencies where the upstream tasks ("predecessors") are not spawned yet
+    and thus making downstream tasks ("successors") wait.
+    """
 
-        When a task corresponding a key is spawned,
-        it iterates a value, the list of the dependee tasks,
-        and notifes them their one dependency is spawned.
+    # A dictionary with predecessors as keys and (lists of) successors as values,
+    # i.e. value tasks are waiting for key tasks.
+    #
+    # Example:
+    #   Predecessor A -> [Successor B, Successor C]
+    #
+    # - When A is spawned, it should notify B and C.
+    # - B and C must have already been spawned.
+    _dependencies: Dict['TaskID', List['TaskID']]
 
-        If all dependencies of a dependee task are spawned,
-        it is also spawned.
-    """
-    _wait_dependees_dict: Dict[TaskID, List[TaskID]]
     def __init__(self):
         self._mutex = threading.Lock()
-        self._wait_dependees_dict = {}
+        self._dependencies = defaultdict(list)
 
-    def append_wait_task(self, dependency: TaskID, dependee: TaskID):
-        """ Append or accumulate a dependee task for a dependency task key """
+    def add(self, predecessor: 'TaskID', successor: 'TaskID'):
+        if successor.task is None:
+            raise ValueError(
+                "successor task %s should have been spawned", repr(successor)
+            )
         with self._mutex:
-            if dependency in self._wait_dependees_dict:
-                self._wait_dependees_dict[dependency].append(dependee)
-            else:
-                self._wait_dependees_dict[dependency] = [dependee]
+            self._dependencies[predecessor].append(successor)
```

Comment:
I'm not sure I like the mixing of `_dependencies` with `predecessor` and `successor`. 

I also don't have better suggestions... 🤔

Reply:
same here. imo, successor -> dependee, predecessor -> dependency looks clearer.

Reply:
plus, does the `append()` create new key if the key does not exist?

Reply:
I understand "dependency" as a relation; "predecessor" and "successor" are roles. "dependee" is confusing.

Reply:
then let's just replace `_dependencies` with `_successors`